### ISSUE TITLE
Compiler fails to compile int8 quantized TFLite Model

### ISSUE THREAD

**Issue by johannes-do** [google-coral/edgetpu#453]
### Description

I want to convert a int8 quantized tflite model:

```python
converter.optimizations = [tf.lite.Optimize.DEFAULT]
def representative_dataset_gen():
  # Get sample input data as a numpy array in a method of your choosing.
  for data in dataset:
    # not sure why, for single input data use [data] for multiple inputs just data...
    yield [data]
converter.representative_dataset = representative_dataset_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
converter.experimental_new_converter = True
tflite_model = converter.convert()
```

I can load the quantized tflite model without issues and run inference with it. But when I want to convert it the edgetpu compiler fails with many of these errors:
```
Edge TPU Compiler version 16.0.384591198
Searching for valid delegate with step 1
Try to compile segment with 136 ops
Started a compilation timeout timer of 180 seconds.
ERROR: :344 no_integer_overflow_from_quantization was not true.
ERROR: Node number 31 (CONV_2D) failed to prepare.

Compilation failed: Model failed in Tflite interpreter. Please ensure model can be loaded/run in Tflite interpreter.
Compilation child process completed within timeout period.
Compilation failed! 
```

I assume this means that somewhere in the quantized model there are integers outside of int8 range. But that should not happen with the quantized tflite model, right?
[model_quant.tflite.zip](https://github.com/google-coral/edgetpu/files/7062227/model_quant.tflite.zip)

And thanks for adding more meaningful error messages with the new compiler update!

<details><summary>Click to expand!</summary> 

### Issue Type

Bug, Support

### Operating System

Linux

### Programming Language

Python 3.7

</details>

**Comment by hjonnala** [comment]
Hi, @j-o-d-o can you please try commenting out this line:

`converter.experimental_new_converter = True`

if problem still persist, can you please share your saved model?

**Comment by johannes-do** [comment]
@hjonnala, tried with both new and old converter, same result

**Comment by johannes-do** [comment]
https://www.dropbox.com/s/4hakehtsmhuepsh/tf_model_4.zip?dl=0

**Comment by hjonnala** [comment]
@j-o-d-o can you try to export the saved_model without batchnorm for layer 31 ? 
[model.html.tar.gz](https://github.com/google-coral/edgetpu/files/7078545/model.html.tar.gz)

**Comment by johannes-do** [comment]
@hjonnala that doesnt seem to help. Same error message for the next Layer:
```
ERROR: :344 no_integer_overflow_from_quantization was not true.
ERROR: Node number 37 (CONV_2D) failed to prepare.
```
I mean, I know what the error is trying to tell me. What I do not understand is why they are happening and how I would look into this issue. I am a bit surprised that all batchnorm and bias layers are quantized with int32 despite setting `converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]` and it seems that might be the issue when trying to compile the edgetpu model.

But when I try to compile the model without any bias and with no batchnorm layers, I still get the same error. When I look into the quantized weights of the respective layer in the error message, I get values between [-127, 127]. So I guess that should be fine.

**Comment by johannes-do** [comment]
Any updates on this one? I am kind of stuck on here :(

**Comment by hjonnala** [comment]
Have you exported the model with fixed tensor shape? only fixed shape tensors or models will work on edgetpu.

 I have tried to quantize the model the saved model and I am getting 
`ERROR: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.`. 

[edgetpu_453.ipynb.tar.gz](https://github.com/google-coral/edgetpu/files/7221037/edgetpu_453.ipynb.tar.gz)

**Comment by xadrianzetx** [comment]
Hi, any updates on this? I'm having the same problem. Converter  setup is:

```python
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
converter.optimizations = [tf.lite.Optimize.DEFAULT]
```

and compiling with `edgetpu_compiler -a -s model.tflite` yields

```
Edge TPU Compiler version 16.0.384591198
Started a compilation timeout timer of 180 seconds.
ERROR: :344 no_integer_overflow_from_quantization was not true.
ERROR: Node number 15 (DEPTHWISE_CONV_2D) failed to prepare.

Compilation failed: Model failed in Tflite interpreter. Please ensure model can be loaded/run in Tflite interpreter.
Compilation child process completed within timeout period.
Compilation failed! 
```

That's puzzling, since the same model compiles fine with compiler version `15.0.340273435` and `14.1.317412892`. [Here's](https://github.com/google-coral/edgetpu/files/7278575/model.tar.gz) the tflite model in question.

**Comment by johannes-do** [comment]
> Have you exported the model with fixed tensor shape? only fixed shape tensors or models will work on edgetpu.
> 
> I have tried to quantize the model the saved model and I am getting `ERROR: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.`.
> 
> [edgetpu_453.ipynb.tar.gz](https://github.com/google-coral/edgetpu/files/7221037/edgetpu_453.ipynb.tar.gz)

@hjonnala This error happens when setting `converter._experimental_new_quantizer = True`. After training I have to train one epoch with a fixed batch size of 1 (instead of 'None') which will resolve this issue. I think it is also possible to fix the batch size to a certain value after training without having to retrain, but my google skills let me down...

So @xadrianzetx, maybe this will help you:
- Make sure to have a fixed batch size of 1. (Greater than 1 can not be mapped to the edgetpu).
- Use `converter._experimental_new_quantizer = True` when doing quantization with tflite

Then it should also work again with the latest edgetpu compiler. But this is all just a workaround for the original issue.

**Comment by hjonnala** [comment]
@xadrianzetx  if you are still having the issue please share your saved model and quantization script.

Thanks

**Comment by johannes-do** [comment]
@hjonnala The original issue I posted is still present. My last message is only a workaround. I don't think it is a viable solution to this issue to tell developers that they must set an unsupported flag (_experimental_new_quantizer). 
Also, if I were to develop this compiler I would really want to know why it fails with the default quantization and works with the experimental_new_quantizer.

**Comment by hjonnala** [comment]
@j-o-d-o  can you please share saved model and quantization script that you were able to compile the model. 

Also, please compare the output quantization at Node number 31. (tflite able to compile vs tflite not able to compile)

**Comment by xadrianzetx** [comment]
@hjonnala 

[This colab notebook](https://colab.research.google.com/drive/1qHBF_L6od9Ck2Y2_v7POfLelrovc4Db0?usp=sharing) contains reproducible example. The same model will compile with older compilers (14.1 and 15.0). Also, when using newer version of tensorflow to create a model, 16.0 will compile it, but depthwise conv layers will run on CPU, similarly to #75.

**Comment by hjonnala** [comment]
@xadrianzetx yes, its due to  depth wise conv layer limitation(Tensor has unsupported rank (up to 3 innermost dimensions mapped), those ops  are mapped to CPU.

**Comment by hjonnala** [comment]
Fee free to reopen if you have questions.

**Comment by estebanvalles** [comment]
I have a similar problem with compilation. Not sure if it should be part of this thread or a separate thread, but it seems to be an issue with weight conversion to int8

My model compiles when I create it **as long as I don't load my Tensorflow weights:** 
- If I create the model, don't train or load weights, quantize and compile with the same settings as in this thread, things work. 
- When I load pre-trained weights, and repeat the procedure (quantize and compile), it doesn't work and fails without an error message:
`edgetpu_compiler -a -s -t 1000 {save_path}` 
returns:
```
Edge TPU Compiler version 16.0.384591198
Started a compilation timeout timer of 1000 seconds.
Compilation child process completed within timeout period
Compilation failed!   
```

When pre-trained weights are not loaded, the log file shows the following layers:
```
Edge TPU Compiler version 16.0.384591198  
Input: /home/ubuntu/network_quant.tflite   
Output: network_quant_edgetpu.tflite  
  
Operator                       Count      Status
CONCATENATION                  3          Mapped to Edge TPU
SOFTMAX                        1          Mapped to Edge TPU
RESHAPE                        2          Mapped to Edge TPU
ADD                            19         Mapped to Edge TPU
QUANTIZE                       2          Mapped to Edge TPU
TRANSPOSE_CONV                 3          Mapped to Edge TPU
CONV_2D                        59         Mapped to Edge TPU
```

Software versions:  Tensorflow 2.8.0, Python 3.9.12
```
converter.optimizations = [tf.lite.Optimize.DEFAULT] 
converter.representative_dataset = representative_data_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8 
```
